### ISSUE TITLE
Unexpected connections count growth

### ISSUE THREAD

**Issue by drakmail** [sorintlab/stolon#634]
### Submission type

- Bug report

### Environment

I have 3 nodes kubernetes cluster with 3 nodes stolon installation:

```
$ kubectl version
Client Version: version.Info{Major:"1", Minor:"13", GitVersion:"v1.13.3", GitCommit:"721bfa751924da8d1680787490c54b9179b1fed0", GitTreeState:"clean", BuildDate:"2019-02-01T20:08:12Z", GoVersion:"go1.11.5", Compiler:"gc", Platform:"linux/amd64"}
Server Version: version.Info{Major:"1", Minor:"13", GitVersion:"v1.13.3", GitCommit:"721bfa751924da8d1680787490c54b9179b1fed0", GitTreeState:"clean", BuildDate:"2019-02-01T20:00:57Z", GoVersion:"go1.11.5", Compiler:"gc", Platform:"linux/amd64"}
```

```
stolon-keeper: sorintlab/stolon:master-pg11
stolon-proxy: sorintlab/stolon:master-pg10 # hm, didn't notice these earlier, maybe need to change to master-pg11?
```

`stolonctl --cluster-name=kube-stolon --store-backend=kubernetes --kube-resource-kind=configmap clusterdata | jq`

```json
{
  "formatVersion": 1,
  "changeTime": "2019-04-17T17:39:34.414999099Z",
  "cluster": {
    "uid": "8dd3f54d",
    "generation": 1,
    "changeTime": "2019-04-17T06:49:41.445925319Z",
    "spec": {
      "additionalWalSenders": null,
      "additionalMasterReplicationSlots": null,
      "initMode": "new",
      "pgParameters": {
        "datestyle": "iso, mdy",
        "default_text_search_config": "pg_catalog.english",
        "dynamic_shared_memory_type": "posix",
        "lc_messages": "en_US.utf8",
        "lc_monetary": "en_US.utf8",
        "lc_numeric": "en_US.utf8",
        "lc_time": "en_US.utf8",
        "log_timezone": "UTC",
        "max_connections": "500",
        "max_wal_size": "1GB",
        "min_wal_size": "80MB",
        "shared_buffers": "128MB",
        "timezone": "UTC",
        "wal_level": "replica"
      },
      "pgHBA": null
    },
    "status": {
      "phase": "normal",
      "master": "950598ce"
    }
  },
  "keepers": {
    "keeper0": {
      "uid": "keeper0",
      "generation": 1,
      "changeTime": "2019-04-17T17:39:34.415101679Z",
      "spec": {},
      "status": {
        "healthy": true,
        "lastHealthyTime": "2019-04-17T17:39:34.412181803Z",
        "bootUUID": "cfec2749-e8ec-4c14-992b-bfb48dea79f8",
        "postgresBinaryVersion": {
          "Maj": 11,
          "Min": 1
        }
      }
    },
    "keeper1": {
      "uid": "keeper1",
      "generation": 1,
      "changeTime": "2019-04-17T17:39:34.415103219Z",
      "spec": {},
      "status": {
        "healthy": true,
        "lastHealthyTime": "2019-04-17T17:39:34.412231739Z",
        "bootUUID": "75f6f0ea-7538-46ec-880c-90f0de2a85c5",
        "postgresBinaryVersion": {
          "Maj": 11,
          "Min": 1
        }
      }
    },
    "keeper2": {
      "uid": "keeper2",
      "generation": 1,
      "changeTime": "2019-04-17T17:39:34.415104572Z",
      "spec": {},
      "status": {
        "healthy": true,
        "lastHealthyTime": "2019-04-17T17:39:34.412231955Z",
        "bootUUID": "03071133-b428-4027-8f56-89c0dc2f724d",
        "postgresBinaryVersion": {
          "Maj": 11,
          "Min": 1
        }
      }
    }
  },
  "dbs": {
    "950598ce": {
      "uid": "950598ce",
      "generation": 4,
      "changeTime": "2019-04-17T17:39:29.063308127Z",
      "spec": {
        "keeperUID": "keeper0",
        "requestTimeout": "10s",
        "maxStandbys": 20,
        "additionalWalSenders": 5,
        "additionalReplicationSlots": null,
        "initMode": "none",
        "pgParameters": {
          "datestyle": "iso, mdy",
          "default_text_search_config": "pg_catalog.english",
          "dynamic_shared_memory_type": "posix",
          "lc_messages": "en_US.utf8",
          "lc_monetary": "en_US.utf8",
          "lc_numeric": "en_US.utf8",
          "lc_time": "en_US.utf8",
          "log_timezone": "UTC",
          "max_connections": "500",
          "max_wal_size": "1GB",
          "min_wal_size": "80MB",
          "shared_buffers": "128MB",
          "timezone": "UTC",
          "wal_level": "replica"
        },
        "pgHBA": null,
        "role": "master",
        "followers": [
          "b35bfbc7",
          "b7c9f2b8"
        ],
        "synchronousStandbys": null,
        "externalSynchronousStandbys": null
      },
      "status": {
        "healthy": true,
        "currentGeneration": 4,
        "listenAddress": "10.46.0.47",
        "port": "5432",
        "systemdID": "6661564963581435952",
        "timelineID": 35,
        "xLogPos": 431962980352,
        "timelinesHistory": [
          {
            "timelineID": 1,
            "switchPoint": 104152544,
            "reason": "no recovery target specified"
          },
          {
            "timelineID": 2,
            "switchPoint": 168116288,
            "reason": "no recovery target specified"
          },
          {
            "timelineID": 3,
            "switchPoint": 241177048,
            "reason": "no recovery target specified"
          },
          {
            "timelineID": 4,
            "switchPoint": 319278176,
            "reason": "no recovery target specified"
          },
          {
            "timelineID": 5,
            "switchPoint": 408242520,
            "reason": "no recovery target specified"
          },
          {
            "timelineID": 6,
            "switchPoint": 438880352,
            "reason": "no recovery target specified"
          },
          {
            "timelineID": 7,
            "switchPoint": 545748896,
            "reason": "no recovery target specified"
          },
          {
            "timelineID": 8,
            "switchPoint": 616559360,
            "reason": "no recovery target specified"
          },
          {
            "timelineID": 9,
            "switchPoint": 661846880,
            "reason": "no recovery target specified"
          },
          {
            "timelineID": 10,
            "switchPoint": 772009920,
            "reason": "no recovery target specified"
          },
          {
            "timelineID": 11,
            "switchPoint": 838860800,
            "reason": "no recovery target specified"
          },
          {
            "timelineID": 12,
            "switchPoint": 889291024,
            "reason": "no recovery target specified"
          },
          {
            "timelineID": 13,
            "switchPoint": 957347112,
            "reason": "no recovery target specified"
          },
          {
            "timelineID": 14,
            "switchPoint": 1029460176,
            "reason": "no recovery target specified"
          },
          {
            "timelineID": 15,
            "switchPoint": 1084824168,
            "reason": "no recovery target specified"
          },
          {
            "timelineID": 16,
            "switchPoint": 1179993424,
            "reason": "no recovery target specified"
          },
          {
            "timelineID": 17,
            "switchPoint": 1251950672,
            "reason": "no recovery target specified"
          },
          {
            "timelineID": 18,
            "switchPoint": 1303772448,
            "reason": "no recovery target specified"
          },
          {
            "timelineID": 19,
            "switchPoint": 1402020456,
            "reason": "no recovery target specified"
          },
          {
            "timelineID": 20,
            "switchPoint": 1433996208,
            "reason": "no recovery target specified"
          },
          {
            "timelineID": 21,
            "switchPoint": 1498995600,
            "reason": "no recovery target specified"
          },
          {
            "timelineID": 22,
            "switchPoint": 1529883184,
            "reason": "no recovery target specified"
          },
          {
            "timelineID": 23,
            "switchPoint": 1594208576,
            "reason": "no recovery target specified"
          },
          {
            "timelineID": 24,
            "switchPoint": 1644316936,
            "reason": "no recovery target specified"
          },
          {
            "timelineID": 25,
            "switchPoint": 1752479056,
            "reason": "no recovery target specified"
          },
          {
            "timelineID": 26,
            "switchPoint": 1833599768,
            "reason": "no recovery target specified"
          },
          {
            "timelineID": 27,
            "switchPoint": 1899577816,
            "reason": "no recovery target specified"
          },
          {
            "timelineID": 28,
            "switchPoint": 1967253712,
            "reason": "no recovery target specified"
          },
          {
            "timelineID": 29,
            "switchPoint": 1996545920,
            "reason": "no recovery target specified"
          },
          {
            "timelineID": 30,
            "switchPoint": 7478730248,
            "reason": "no recovery target specified"
          },
          {
            "timelineID": 31,
            "switchPoint": 313924290208,
            "reason": "no recovery target specified"
          },
          {
            "timelineID": 32,
            "switchPoint": 341866311936,
            "reason": "no recovery target specified"
          },
          {
            "timelineID": 33,
            "switchPoint": 348711309488,
            "reason": "no recovery target specified"
          },
          {
            "timelineID": 34,
            "switchPoint": 424899771752,
            "reason": "no recovery target specified"
          }
        ],
        "pgParameters": {
          "datestyle": "iso, mdy",
          "default_text_search_config": "pg_catalog.english",
          "dynamic_shared_memory_type": "posix",
          "lc_messages": "en_US.utf8",
          "lc_monetary": "en_US.utf8",
          "lc_numeric": "en_US.utf8",
          "lc_time": "en_US.utf8",
          "log_timezone": "UTC",
          "max_connections": "500",
          "max_wal_size": "1GB",
          "min_wal_size": "80MB",
          "shared_buffers": "128MB",
          "timezone": "UTC",
          "wal_level": "replica"
        },
        "synchronousStandbys": null,
        "olderWalFile": "00000023000000640000008A"
      }
    },
    "b35bfbc7": {
      "uid": "b35bfbc7",
      "generation": 4,
      "changeTime": "2019-04-17T06:50:07.822377643Z",
      "spec": {
        "keeperUID": "keeper2",
        "requestTimeout": "10s",
        "maxStandbys": 20,
        "additionalWalSenders": 5,
        "additionalReplicationSlots": null,
        "initMode": "resync",
        "pgParameters": {
          "datestyle": "iso, mdy",
          "default_text_search_config": "pg_catalog.english",
          "dynamic_shared_memory_type": "posix",
          "lc_messages": "en_US.utf8",
          "lc_monetary": "en_US.utf8",
          "lc_numeric": "en_US.utf8",
          "lc_time": "en_US.utf8",
          "log_timezone": "UTC",
          "max_connections": "500",
          "max_wal_size": "1GB",
          "min_wal_size": "80MB",
          "shared_buffers": "128MB",
          "timezone": "UTC",
          "wal_level": "replica"
        },
        "pgHBA": null,
        "role": "standby",
        "followConfig": {
          "type": "internal",
          "dbuid": "950598ce"
        },
        "followers": [],
        "synchronousStandbys": null,
        "externalSynchronousStandbys": null
      },
      "status": {
        "listenAddress": "10.32.0.9",
        "port": "5432",
        "synchronousStandbys": null
      }
    },
    "b7c9f2b8": {
      "uid": "b7c9f2b8",
      "generation": 1,
      "changeTime": "2019-04-17T06:50:34.301049216Z",
      "spec": {
        "keeperUID": "keeper1",
        "requestTimeout": "10s",
        "maxStandbys": 20,
        "additionalWalSenders": 5,
        "additionalReplicationSlots": null,
        "initMode": "resync",
        "pgParameters": {
          "datestyle": "iso, mdy",
          "default_text_search_config": "pg_catalog.english",
          "dynamic_shared_memory_type": "posix",
          "lc_messages": "en_US.utf8",
          "lc_monetary": "en_US.utf8",
          "lc_numeric": "en_US.utf8",
          "lc_time": "en_US.utf8",
          "log_timezone": "UTC",
          "max_connections": "500",
          "max_wal_size": "1GB",
          "min_wal_size": "80MB",
          "shared_buffers": "128MB",
          "timezone": "UTC",
          "wal_level": "replica"
        },
        "pgHBA": null,
        "role": "standby",
        "followConfig": {
          "type": "internal",
          "dbuid": "950598ce"
        },
        "followers": [],
        "synchronousStandbys": null,
        "externalSynchronousStandbys": null
      },
      "status": {
        "listenAddress": "10.40.0.10",
        "port": "5432",
        "synchronousStandbys": null
      }
    }
  },
  "proxy": {
    "generation": 270,
    "changeTime": "2019-04-17T06:49:46.576514397Z",
    "spec": {
      "masterDbUid": "950598ce",
      "enabledProxies": [
        "5a97af16",
        "c3737b29",
        "d8c2924a"
      ]
    },
    "status": {}
  }
}
```

```
stolonctl --cluster-name=kube-stolon --store-backend=kubernetes --kube-resource-kind=configmap status
=== Active sentinels ===

ID              LEADER
137504db        true
2a04bd06        false
fa1683ab        false

=== Active proxies ===

ID
5a97af16
c3737b29
d8c2924a

=== Keepers ===

UID     HEALTHY PG LISTENADDRESS        PG HEALTHY      PG WANTEDGENERATION     PG CURRENTGENERATION
keeper0 true    10.46.0.47:5432         true            4                       4
keeper1 true    10.40.0.10:5432         false           1                       0
keeper2 true    10.32.0.9:5432          false           4                       0

=== Cluster Info ===

Master: keeper0

===== Keepers/DB tree =====

keeper0 (master)
├─keeper2
└─keeper1
```

#### Stolon version

`v0.13.0`

### Unexpected behaviour you saw

After several days stolon user connections count has growth to about 50:

```
postgres=# select client_addr, usename, count(*) from pg_stat_activity group by usename, client_addr;
 client_addr |   usename    | count 
-------------+--------------+-------
             |              |     4
 10.32.0.10  | censored     |     9
 10.40.0.10  | repluser     |     1
 10.40.0.17  | censored     |    12
 10.46.0.53  | sentry       |     1
             | stolon       |    56
 10.40.0.17  | censored     |    22
 10.46.0.53  | censored     |    12
 10.40.0.17  | sentry       |     1
 10.32.0.10  | stolon       |     1
 10.32.0.9   | repluser     |     1
 10.32.0.10  | censored     |    25
 10.46.0.53  | censored     |    21
(13 rows)
```

Some time ago after running stolon for about a month I had has more than 400 opened connections from stolon. After killing 75% of connections nothing bad happened and stolon continue working.

### Steps to reproduce the problem

In my case I need just to wait...

**Comment by sgotti** [comment]
@drakmail Thanks for your report. Can you please try with #635 ? It was in my queue for some time but didn't had time to create a PR.

**Comment by drakmail** [comment]
@sgotti to try it do I need just to build custom docker image based on #635 and update k8s manifest?

**Comment by sgotti** [comment]
@drakmail hopefully fixed by #635. Please ask to reopen if this happens again. Thanks!
